Source/Orts.Simulation/Simulation/Signalling/CsSignalScript.cs

```diff
@@ -0,0 +1,375 @@
+        /// Distance at which the signal will be cleared during approach control
+        /// </summary>
+        public float? ApproachControlRequiredPosition => SignalHead.ApproachControlLimitPositionM;
+        /// <summary>
+        /// Maximum train speed at which the signal will be cleared during approach control
+        /// </summary>
+        public float? ApproachControlRequiredSpeed => SignalHead.ApproachControlLimitSpeedMpS;
+        /// <summary>
+        /// Occupation and reservation state of the signal's route
+        /// </summary>
+        public BlockState CurrentBlockState => (BlockState)SignalObject.block_state();
+        /// <summary>
+        /// True if the signal link is activated
+        /// </summary>
+        public bool RouteSet => SignalHead.route_set() > 0;
+        /// <summary>
+        /// Set this variable to true to allow clear to partial route
+        /// </summary>
+        public bool AllowClearPartialRoute { set { SignalObject.AllowClearPartialRoute(value ? 1 : 0); } }
+        /// <summary>
+        /// Number of signals to clear ahead of this signal
+        /// </summary>
+        public int SignalNumClearAhead { get { return SignalObject.SignalNumClearAheadActive; } set { SignalObject.SetSignalNumClearAhead(value); } }
+        /// <summary>
+        /// Default draw state of the signal for a specific aspect
+        /// </summary>
+        /// <param name="signalAspect">Aspect for which the default draw state must be found</param>
+        /// <returns></returns>
+        public int DefaultDrawState(Aspect signalAspect) => SignalHead.def_draw_state((MstsSignalAspect)signalAspect);
+        /// <summary>
+        /// Index of the draw state with the specified name
+        /// </summary>
+        /// <param name="name">Name of the draw state as defined in sigcfg</param>
+        /// <returns>The index of the draw state, -1 if no one exist with that name</returns>
+        public int GetDrawState(string name) => SignalHead.signalType.DrawStates.TryGetValue(name, out SignalDrawState drawState) ? drawState.Index : -1;
+        /// <summary>
+        /// Signal identity of this signal
+        /// </summary>
+        public int SignalId => SignalObject.thisRef;
+        /// <summary>
+        /// Name of this signal type, as defined in sigcfg
+        /// </summary>
+        public string SignalTypeName => SignalHead.SignalTypeName;
+        /// <summary>
+        /// Local storage of this signal, which can be accessed from other signals
+        /// </summary>
+        public Dictionary<int, int> SharedVariables => SignalObject.localStorage;
+        /// <summary>
+        /// Current time in seconds
+        /// </summary>
+        public float ClockTimeS => (float)SignalObject.signalRef.Simulator.GameTime;
+        /// <summary>
+        /// Represents a timer which can be started and stopped several times, with a configurable trigger value
+        /// </summary>
+        public class Timer
```

Comment:
Ideally we wouldn't have to copy and paste this code for every new scriptable subsystem. I'll make a note to investigate whether we can share it among multiple classes.

Reply:
One possibility is that CsSignalScript inherits from AbstractScriptClass, where these are defined. Then we would have to create a new abstract class for train scripts, adding DistanceM(), SpeedMpS() and Odometer functionality there.

What do you think about using the Orts.Scripting.Script namespace instead of Orts.Signalling.Signals for the signal scripts?

Reply:
That sounds like a good idea. I agree it would be more fitting to place a scriptable subsystem under the Script namespace.

Reply:
Now the signal scripts inherit from AbstractScriptClass, so the built-in timer can be used, They must reside in the ORTS.Scripting.Script namespace to be loaded.